packages/backend/src/util/handlers/createExpressSynthqlHandler.ts

```diff
@@ -17,47 +17,129 @@ export function createExpressSynthqlHandler<T>(
-                        queryEngine.execute(query, {
-                            returnLastOnly,
-                        }),
-                    );
-
-                    res.statusCode = 200;
-                    res.setHeader('Content-Type', 'application/json');
-                    res.write(JSON.stringify(result));
-                    res.end();
-                } catch (error) {
-                    res.statusCode = 500;
-                    res.setHeader('Content-Type', 'application/json');
-                    res.write(JSON.stringify({ error: String(error) }));
-                    res.end();
-                }
+            await executeSynthqlRequest<T>(queryEngine, req, res);
+        } catch (e) {
+            // Handle known `SynthqlError`s
+            if (e instanceof SynthqlError) {
+                res.status(400).json({
+                    type: e.type,
+                    error: e.message,
+                });
             } else {
-                res.statusCode = 200;
-                res.setHeader('Content-Type', 'application/x-ndjson');
+                // Let another layer handle the error
+                throw e;
+            }
+        }
+    };
+}
+
+async function executeSynthqlRequest<T>(
+    queryEngine: QueryEngine<T>,
+    req: ExpressSynthqlHandlerRequest,
+    res: ExpressSynthqlHandlerResponse,
+) {
+    // First try to parse the request body as JSON
+    const { query, returnLastOnly } = await tryParseRequest(req);
+
+    // We don't do this yet, but eventually we'll want to validate the request
+    // const validatedQuery = await tryValidateSynthqlQuery(query);
+
+    // Execute the query, but just to get the initial generator
+    const resultGenerator = await tryExecuteQuery<T>(
+        queryEngine,
+        query,
+        returnLastOnly,
+    );
+
+    // Now that we have the generator, we want to iterate over the items
+    // and depending on `returnLastOnly`, we will write the status code
+    // either before, or after iteration
+    await writeBody(res, query, resultGenerator, returnLastOnly);
+
+    res.end();
+}
+
+async function tryParseRequest(req: ExpressSynthqlHandlerRequest) {
+    const body = req.body;
+    const returnLastOnly = req.headers['x-return-last-only'] === 'true';
 
-                for await (const intermediateResult of queryEngine.execute(
-                    query,
-                )) {
-                    res.write(JSON.stringify(intermediateResult));
-                    res.write('\n');
-                }
+    try {
+        const query = JSON.parse(body);
 
-                res.end();
+        return { query, returnLastOnly };
+    } catch (e) {
+        throw SynthqlError.createJsonParsingError({
+            error: e,
+            json: body,
+        });
+    }
+}
+
+async function tryExecuteQuery<T>(
+    queryEngine: QueryEngine<T>,
+    query: any,
+    returnLastOnly: boolean,
+) {
+    if (returnLastOnly) {
```

Comment:
isn't this whole function just `queryEngine.execute(query, {returnLastOnly})`? no need for an `if`

Reply:
It should be, but at the time I wrote this, I wasn't sure how passing a falsy value to returnLastOnly was being used inside. After looking inside yesterday, I can see why this is unnecessary. Will fix now